cardano-crypto-class/src/Cardano/Crypto/KES/Class.hs

```diff
@@ -277,6 +284,46 @@ updateKES
 updateKES = updateKESWith mlockedMalloc
 
 
+-- | Pure implementations of the core KES operations. These are unsound, because
+-- proper handling of KES secrets (seeds, sign keys) requires mlocking and
+-- deterministic erasure (\"secure forgetting\"), which is not possible in pure
+-- code.
+-- This API is only provided for testing purposes; it must not be used to
+-- generate or use real KES keys.
+class KESAlgorithm v => UnsoundPureKESAlgorithm v where
+  data UnsoundPureSignKeyKES v :: Type
+
+  unsoundPureSignKES
+    :: forall a. (Signable v a)
+    => ContextKES v
+    -> Period  -- ^ The /current/ period for the key
+    -> a
+    -> UnsoundPureSignKeyKES v
+    -> SigKES v
+
+  unsoundPureUpdateKES
+    :: ContextKES v
+    -> UnsoundPureSignKeyKES v
+    -> Period  -- ^ The /current/ period for the key, not the target period.
+    -> Maybe (UnsoundPureSignKeyKES v)
+
+  unsoundPureGenKeyKES
+    :: Seed
+    -> UnsoundPureSignKeyKES v
+
+  unsoundPureDeriveVerKeyKES
+    :: UnsoundPureSignKeyKES v
+    -> VerKeyKES v
+
+  unsoundPureSignKeyKESToSoundSignKeyKES
+    :: (MonadST m, MonadThrow m)
+    => UnsoundPureSignKeyKES v
+    -> m (SignKeyKES v)
```

Comment:
IIRC, the thought behind this was that this allows for a more efficient implementation for `MockKES` e.a. (which don't actually mlock anything, and just use the same data structures for both sound and unsound keys, modulo newtypes):

```haskell
    unsoundPureSignKeyKESToSoundSignKeyKES (UnsoundPureSignKeyMockKES vk t) =
      return $ SignKeyMockKES vk t
```

(Come to think of it, this could possibly even be boiled down to just `unsoundPureSignKeyKESToSoundSignKeyKES = coerce`, though I'm not sure whether GHC is smart enough to allow this.)

Anyway, the point is that the raw serialization round trip is an implementation detail. We could also extract the seed and then `genKeyKES`, or we could manually allocate a buffer and then `memcopy` the key over; it's just that raw serialization happens to be efficient enough, and the safest and most convenient implementation.

We could provide this version as a default implementation though.

Reply:
I wouldn't worry about efficiency for `MockKES`, since it is only ever used for testing

Reply:
Right. Although in this particular case, efficiency is many times more relevant for testing than in production use. In production, we will do this conversion once when the node kernel (re)starts, but in the test suite, we exercise it thousands of times. The difference is probably not huge, but considering that part of the point of `MockKES` is to avoid doing more work than necessary in tests, I don't think enshrining serialization as the only method for doing this conversion would be a good idea.

Reply:
Oh, and there's another thing - for SumKES, the sign key contains not just the actual private key, but also a pair of verification keys (and this stacks, because the higher-order SumKES are defined recursively, so Sum5KES contains 10 verification keys). We can do the marshalling through ser/deser, but that's wasteful, because the verkeys aren't mlocked, so we can just share them rather than making a deep copy; the only thing we need to actually copy are the seeds, one per level (so that's 5 seeds for Sum5KES), cutting the amount of copying down by 2/3.

Reply:
I don't really have a strong opinion on this topic. If you think it is worthwhile to keep it a type class method, by all means. I just don't think speculative improvement of performance of the test suite is a good reason, but I am not gonna oppose it either.

> we can just share them rather than making a deep copy; the only thing we need to actually copy are the seeds, one per level (so that's 5 seeds for Sum5KES), cutting the amount of copying down by 2/3.

Please, create a ticket about it and let's not do any further optimizations in this PR.

I really want this work to get merged, so we can finally make a very much overdue release of `cardano-crypto-class`.

Reply:
The "optimization" is already in place, and has been the entire time. I'll keep the code as-is for now if that's OK.